### DISCUSSION TITLE
Propagating Exceptions in Interactive Shell

### DISCUSSION THREAD

**Discussion by irinaleo** [remkop/picocli#2055] [Q&A] [answered]
Hello,

I am following along with [this ](https://github.com/remkop/picocli/wiki/JLine-3-Examples#example-for-jline-3141-and-picocli-430---432)example adding picoCLI to an existing interactive shell. It has some existing Spock tests that look for expected thrown Exceptions in the business logic.

 I believe CommandLine::execute is swallowing the Exceptions, but I want to preserve my tests.

After much digging, I think the best thing to do is implement [DIY command execution](https://picocli.info/#_diy_command_execution).

Since this is a shell, I wanted to know if there are any tips or pointers you can give that could save me some time e.g. is what I'm thinking even possible?

(Also, If possible, it would be helpful to add a flag to allow Execution propagation or something along those lines for future users.)

**Comment by remkop** [comment, marked as answer]
The `execute` method catches all exceptions, that is the intention.

One idea is to have [custom exception handlers](https://picocli.info/#_handling_errors):
Generally there are 2 types of errors: invalid user input and the application throwing an exception in its business logic.
There are separate `IParameterExceptionHandler` and `IExecutionExceptionHandler` interfaces for these.

You could have a custom `IExecutionExceptionHandler` that preserves the exception thrown by the business logic (in addition to printing a stack trace and returning a non-zero error code). The test code could then inspect whether the caught and preserved exception matches the expectation.

Alternatively you could do "black box" testing and inspect the Standard Error stream to see if a stack trace was printed, and if it contains the name of the expected exception class.

**Reply by irinaleo** [reply]
I see. That makes sense why I could see the exception from my custom handler but not when I re-threw it from there.
I like the black box testing approach, so I'll move forward with that. Thank you for your insight.
